Eugene branch

sizov opened the pull request.

sizov committed:
Adding readme.md

sizov committed:
Adding gitignore

sizov committed:
Adding gitignore

sizov committed:
Adding gitignore

sizov committed:
Adding gitignore

sizov committed:
Adding gitignore

sizov committed:
Adding gitignore

sizov committed:
Fixing gitignore

sizov committed:
Fixing gitignore

sizov committed:
Merge remote-tracking branch 'remotes/origin/release/0.0.1' into eugene-branch

sizov committed:
Adding files back

sizov committed:
Simplifying gitignore

sizov committed:
Removing gitignore

sizov commented:
Hello Eugene, I have created branch for your developent called 'eugene-branch'. This pull request is to show all the files that you've added. I will leave comments now on the places that I think need work.
Also, please try installing WebStorm - it will help you to deal with formatting your code.

sizov commented on index.html:
```diff
@@ -0,0 +1,25 @@
+  This is a page
+  a simple page
+
+<div id="outputF">
```
It is better to name variables with meaningful name. For example `id='outputContainer'` instead of `id='outputF'`

sizov commented on myscript.js:
```diff
@@ -0,0 +1,51 @@
+var outputArr = [];
+var outputLine = 0;
+var outputF = function (outputItem) {
+    outputArr [outputLine]=(outputItem);
```
two comments here:
- you don't need space between array variable name and square brackets
- no need to add brackets around `outputItem`, so code above becomes:

`outputArr[outputLine] = outputItem;`

sizov commented on myscript.js:
```diff
@@ -0,0 +1,51 @@
+var outputF = function (outputItem) {
+    outputArr [outputLine]=(outputItem);
+    outputLine = outputLine + 1;
+    $("#outputF").text(outputArr.join(" - "));
```
you project does not have jquery installed, please add it to the project (as simple file)

evgenevolkov committed:
TODO list is completely done.

evgenevolkov committed:
TODO list is completely done.

sizov commented on myscript.js:
```diff
@@ -0,0 +1,21 @@
+
+var dataArray = [1, 8, 7, [ 4, 65, 1, [ 2, 7 ], 45, [ 6, 4], 6 ], 4, 8]
+
+var jenyaFlatternArray = function (Array) {
```
There is a convention to have name of variables starting with lower case, like this `myVariable`.
Then classes should be named with capital letter, like this:

```
var Person = function(name) {};
var me = new Person('Vova')
```

Also `Array` is reserved word for `Array` class, so let's change it to `inputArray`

evgenevolkov replied:
Sure, it's my fault to name variable starting with upper case. I already know it but didn't pay enough attention. 
Also, can You recommend me where can I read more about ''traditional code writing rules" like using ' instead of ", spacing etc.

sizov replied:
sure, I think this is very good guideline: https://github.com/airbnb/javascript
some things might not be clear for you, but stick to the rules that you can understand for now

sizov commented on myscript.js:
```diff
@@ -0,0 +1,21 @@
+var jenyaFlatternArray = function (Array) {
+    var flattering = function (mlArray) {
+        for (var i = 0; i < mlArray.length; i++) {
+            var check = typeof mlArray[i];
```
no need to have local variable `check`, just do `if (typeof mlArray)`

sizov replied:
Also, you see that you have reused line `mlArray[i]` 3 times, so let's store it in local variable and use it instead of `mlArray[i]`

evgenevolkov replied:
Concerning additional local variable `check`, it just didn't work in jsfiddle when I wrote `if (typeof mlArray[i] === "object")`. I couldn't find out the reason,  so I used that needless local variable.
Currently, I changed it as you wrote and it works just fine.

I've added a local variable `var mlArrayI = mlArray[i];`. 
I suppose it should be done so because of the rule of code writing called DRY ('don't repeat yourself'), but can You please explain me more detailed what is the advantage to have another line of code and another one local variable? 
I may suggest it is because it is easier to make a "link" for the computer to an exact place in memory at the address of `mlArray[i]` and later accessing it directly instead of "calculating" its place each time. But it seems weird for me as `mlArray[i]' looks _direct_ enough. 
It will be nice to have a short talk with you about this concern.

sizov replied:
wow, brilliant mention of 'DRY'!!!!! you are absolutely right! believe me, authors of some huge applicates don't know this rule

sizov replied:
> but can You please explain me more detailed what is the advantage to have another line of code and another one local variable? 

Sure, there are at least two reasons. 

1) performance - if you do something more than `myArray[i]`, for example you repeat this: `calculateFactorial(666)` - so this calculation is big performance hit and to save processing time - we store result in local variable and then reuse it everywhere

2) in our case performance is not a big concern, BUT referring `mlArray[i]`three times makes it hard to read... I need to check if each three times have same code, for example to make sure that one of them does not have `mlArray[j]` instead of `mlArray[i]`. So request to change that piece is purely for DRY fix

sizov commented:
Generally, very good! Nice job

evgenevolkov committed:
All Vlad proposals up to 25feb2015 implemented.
Code is working correctly.

 BTW why You proposed to use recursion instead of my way to solve the task? Is it supposed to be a *proper* way to do it?
